compiler/test/compilable/test21359.sh

```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env bash
+
+objfile="${OUTPUT_BASE}${OBJ}"
+$DMD -c -m${MODEL} -allinst -of${objfile} ${EXTRA_FILES}/${TEST_NAME}a.d -I${EXTRA_FILES}
+
+# This test is brittle and might break in the future.
```

Comment:
How about checking for duplicate symbols in general:
```
! nm -j bla.o | uniq -d | read
```
worked for the test case.

Reply:
> I wanted to check that the symbol we are looking for is emitted once at least.

Ok.

> Of course other fail_compilation tests now fail, with uglier 'forward reference' errors coming from the glue layer now, without a Loc. :(

These are messages from the glue layer. I guess you can avoid them if the nextOf().merge().deco check is only added for case Taarray.

Reply:
> but that special case is obviously ugly...

I think it would be ok if implemented as a `cast Tfunction` instead of extra checks in the generic case.

Reply:
I guess it could be beneficial to move the `if (type.deco) return type;` before the `switch` though - the switch seems to be exclusively reserved for skipping the merge and returning `type` as-is.

Reply:
Ah, sorry, missed that.

Then my suggestion to only change the AA case could be an option:
```D
        case Taarray:
            if (!type.isTypeAArray().index.merge().deco)
                return type;
            if (!type.nextOf().merge().deco)
                return type;
            break;
```
as I see deco being used in a lot of places to check for resolved semantic. More regressions looming...

The bad types come from `makeNakedAssociativeArray`. `substWildTo` doesn't return a merged type in some cases. Adding `.merge()` in `makeNakedAssociativeArray` helps bla.d, but probably fixing it in `substWildTo` is better:

https://github.com/dlang/dmd/blob/8d90d93c00f3bca8589b84c750ac85b2d87ad1a5/compiler/src/dmd/typesem.d#L7349
shoud return `t.merge()`.

Reply:
Thx for digging further!

> but probably fixing it in `substWildTo` is better

Yeah I guess there we also don't need a special case for TypeFunctions (no qualified TypeFunctions I hope, only qualified function pointers/delegates). After a superficial glance, those seem to be an ugly special case, e.g., there are 3 occurrences of this comment in that file: https://github.com/dlang/dmd/blob/8d90d93c00f3bca8589b84c750ac85b2d87ad1a5/compiler/src/dmd/typesem.d#L2748-L2753